### ISSUE TITLE
Calling `reset()` in onSubmit results in inconsistent state

### ISSUE THREAD

**Issue by ThiefMaster** [final-form/final-form#142]
https://codesandbox.io/s/mzm3m87vvx

When I `reset()` at the end of the (async) onSubmit function, the fields are reset, but the submitSucceeded flag is not reset. When calling reset() at some other point (e.g. the reset button, or even with in a zero-second setTimeout started in onSubmit), it works fine.

**Comment by ThiefMaster** [comment]
Did some digging in the code, and the "complete" callback that sets the submitted state runs only after the onSubmit promise was fully resolved.

Still looking for a proper way to reset from there. FYI, my usecase is that after submitting I hide the form and show success message - but if someone then goes to the next item to act on the form will show up again and needs to be in a completely clean state (as if it was never submitted before).

**Comment by erikras** [comment]
Well, it's a little gross, but you could get around it by changing:
```jsx
reset()
```
to
```jsx
setTimeout(reset)
```

**Comment by ThiefMaster** [comment]
yeah i mentioned that as a workaround in my initial comment, but it feels rather dirty :)

**Comment by samsch** [comment]
Don't re-render the form. Just render it. Like, don't "hide" the form with css, actually drop it from render and then render it again when you need it.

**Comment by ThiefMaster** [comment]
The problem is that I'm using this in a SUI Modal which is wrapped in my FinalForm component. So the FinalForm stays there even when the actual form gets hidden: https://github.com/ThiefMaster/indico/blob/final-form/indico/web/client/js/react/errors/component.jsx#L144

**Comment by erikras** [comment]
One thing I notice about your code is that you should be using the prop `render` rather than `component` for your `modal` function. The `component` prop is going to cause a new react element to be created on every render. That shouldn't be affecting this issue.

Maybe you should use some sort of `setState()` solution? It's not obvious to me that the library should support this particular use case. What do you think?

**Comment by ThiefMaster** [comment]
Would it at least make sense to fail explicitly when trying to call reset() from a place where it will not work as expected?

FWIW, I managed to work around the issue by resetting the form when clicking the "Dismiss" button instead of right after submit. Feels a bit dirty as well but does the job.

**Comment by ThiefMaster** [comment]
Exactly!

**Comment by csantos1113** [comment]
Any updated on this one?

**Comment by erikras** [comment]
I think #221 might have fixed this.

Have you tried since [`v4.13.1`](https://github.com/final-form/final-form/releases/tag/v4.13.1)?

**Comment by csantos1113** [comment]
@erikras - I'm using latest packages, but this error is still present
```
"final-form": "^4.14.1",
"react-final-form": "^6.1.0",
```

See: https://codesandbox.io/s/react-final-form-simple-example-cq7k9

**Comment by erikras** [comment]
Published fix in [`v4.16.0`](https://github.com/final-form/final-form/releases/tag/v4.16.0).
